### ISSUE TITLE
docker with userns doesn't work on aws

### ISSUE THREAD

**Issue by euank** [coreos/bugs#1826]
# Issue Report

### Container Linux Version ###

```
$ cat /etc/os-release
NAME="Container Linux by CoreOS"
ID=coreos
VERSION=1235.12.0
VERSION_ID=1235.12.0
BUILD_ID=2017-02-23-0222
PRETTY_NAME="Container Linux by CoreOS 1235.12.0 (Ladybug)"
ANSI_COLOR="38;5;75"
HOME_URL="https://coreos.com/"
BUG_REPORT_URL="https://github.com/coreos/bugs/issues"
```

### Environment ###

AWS and GCE

### Expected Behavior ###

usernamespaces work identically well on both

### Actual Behavior ###

The EC2 AMI has trouble setting up the rootfs while using userns, the GCE image works fine.

### Reproduction Steps ###

1. Launch a machine using the userdata: `{"ignition":{"version":"2.0.0","config":{}},"storage":{"files":[{"filesystem":"root","path":"/etc/subuid","contents":{"source":"data:,dockremap%3A100000%3A65536","verification":{}},"user":{},"group":{}},{"filesystem":"root","path":"/etc/subgid","contents":{"source":"data:,dockremap%3A100000%3A65536","verification":{}},"user":{},"group":{}}]},"systemd":{"units":[{"name":"docker.service","enable":true,"dropins":[{"name":"10-uesrns.conf","contents":"[Service]\nEnvironment=DOCKER_OPTS=--userns-remap=dockremap"}]}]},"networkd":{},"passwd":{"users":[{"name":"dockremap","create":{}}]}}`
2. `docker run busybox`

Note, a simple reproduction for stracing and such can be achieved with the following instead:
```shell
$ mkdir /tmp/rootfs; cd /tmp/rootfs
$ sudo chown 100000:100000 /tmp/rootfs
$ cat > config.json <<EOF
{
  "ociVersion": "1.0.0-rc2-dev",
  "platform": {
    "os": "linux",
    "arch": "amd64"
  },
  "process": {
    "terminal": true,
    "consoleSize": {
      "height": 0,
      "width": 0
    },
    "user": {
      "uid": 0,
      "gid": 0,
      "additionalGids": [
        10
      ]
    },
    "args": [
      "sh"
    ],
    "env": [
      "PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin",
      "TERM=xterm"
    ],
    "cwd": "/",
    "capabilities": [
      "CAP_CHOWN",
      "CAP_DAC_OVERRIDE",
      "CAP_FSETID",
      "CAP_FOWNER",
      "CAP_MKNOD",
      "CAP_NET_RAW",
      "CAP_SETGID",
      "CAP_SETUID",
      "CAP_SETFCAP",
      "CAP_SETPCAP",
      "CAP_NET_BIND_SERVICE",
      "CAP_SYS_CHROOT",
      "CAP_KILL",
      "CAP_AUDIT_WRITE"
    ]
  },
  "mounts": [
    {
      "destination": "/proc",
      "type": "proc",
      "source": "proc",
      "options": [
        "nosuid",
        "noexec",
        "nodev"
      ]
    }
  ],
  "linux": {
    "uidMappings": [
      {
        "hostID": 100000,
        "containerID": 0,
        "size": 65536
      }
    ],
    "gidMappings": [
      {
        "hostID": 100000,
        "containerID": 0,
        "size": 65536
      }
    ],
    "resources": {
      "disableOOMKiller": false,
      "oomScoreAdj": 0,
      "cpu": {},
      "pids": {
        "limit": 0
      },
      "blockIO": {
        "blkioWeight": 0
      }
    },
    "cgroupsPath": "/runc/123",
    "namespaces": [
      {
        "type": "mount"
      },
      {
        "type": "network"
      },
      {
        "type": "uts"
      },
      {
        "type": "pid"
      },
      {
        "type": "ipc"
      },
      {
        "type": "user"
      }
    ]
  }
}
EOF

$ sudo runc run 123
container_linux.go:247: starting container process caused "process_linux.go:359: container init caused \"rootfs_linux.go:53: mounting \\\"proc\\\" to rootfs \\\"/tmp/root\\\" at \\\"/proc\\\" caused \\\"operation not permitted\\\"\""

# Or on GCE
$ sudo runc run 123
container_linux.go:247: starting container process caused "exec: \"sh\": executable file not found in $PATH"
# success enough
```

### Other Information ###

I'm investigating this, but wanted to make sure it's tracked here in case anyone else is running into it or has an idea what to look at.

Images I'm using: AWS: "CoreOS-stable-1235.12.0-hvm (ami-444dcd24)", GCE: "CoreOS stable 1235.12.0 "

**Comment by euank** [comment]
The simplest repro of this:

```shell
$ cd /tmp
$ unshare -f -mUrpn -- sh -c 'mkdir p2; mount -t proc proc ./p2'; echo $?
# On aws
mount: permission denied
32
# On gce: 
0
```

**Comment by euank** [comment]
And you can fix it with `sudo umount /proc/xen`. That took far longer to notice than it should have.

Docker issue: https://github.com/docker/docker/issues/26070

**Comment by euank** [comment]
Oh, hey, we ship a file that causes this to happen. Of course that's not just magically mounting.

```ini
# /usr/lib/systemd/system/proc-xen.mount
[Unit]
Before=local-fs.target
DefaultDependencies=false
ConditionVirtualization=xen

[Mount]
What=xenfs
Where=/proc/xen
Type=xenfs
```

**Comment by marineam** [comment]
That mount is required for accessing metadata in xenstore like on XenServer and Rackspace.

**Comment by marineam** [comment]
Also, does this come up with any other filesystem that mount under /proc? xenfs certainly isn't the only one.

**Comment by euank** [comment]
@marineam it doesn't come up with any others. I assume that other mounts we have are user-namespace aware, and xenfs's kernel module isn't.

**Comment by euank** [comment]
The patch that should fix this: https://github.com/torvalds/linux/commit/f97df70b1c879f764f88b25b0e67b03a5213968a

That patch is included in 4.10; I'll double check it indeed addresses the issue.

**Comment by euank** [comment]
Verified alpha + userns on aws works fine.
